enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,728 @@
+* Routes with hosts that match the `domainPatterns` of the HSTS will be validated
+* Routes with hosts that match the `domainPatterns` and `namespaceSelector` of the HSTS will be validated
+* Routes that only match the `namespaceSelector` will not be validated
+* Routes that are validated will validate that the `maxAge` exists and falls within the range of the HSTS `maxAgePolicy`
+* Routes that are validated will validate that the `preloadPolicy` and `includeSubdomainPolicy` match the HSTS Policy if they exist
+
+#### Audit HSTS configurations on any cluster
+This proposal offers the cluster administrator a way to review the HSTS configurations guaranteed by the admission plugin.
+For example, to review the maxAge set for required HSTS Policies:
+```shell
+$ oc get clusteroperator/ingress -n openshift-ingress-operator -o jsonpath='{range .spec.requiredHSTSPolicies}{.spec.requiredHSTSPolicies.maxAgePolicy.largestMaxAge}{"\n"}{end}'
+```
+To review the HSTS annotations on all routes:
+````shell
+$ oc get route  --all-namespaces -o go-template='{{range .items}}{{if .metadata.annotations}}{{$a := index .metadata.annotations "haproxy.router.openshift.io/hsts_header"}}{{$n := .metadata.name}}{{with $a}}Name: {{$n}} HSTS: {{$a}}{{"\n"}}{{else}}{{""}}{{end}}{{end}}{{end}}'
+
+Name: myBigRoute HSTS: max-age=31536000;preload;includeSubdomains
+````
+#### Predict HSTS enforcement on any cluster with a single configuration manifest
+This proposal offers the cluster administrator a way to configure and enforce HSTS configurations on routes by configuring
+the `Ingress` using templates that substitute only the for the difference between clusters.  For example, for
+a development cluster with domain `dev.abc.com` and production cluster with `prod.abc.com` domains, administrators can use a single
+`Ingress` configuration template that substitutes the domain:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: abc.com
+  requiredHSTSPolicies:
+    - domainPatterns:
+      - ${DOMAIN}
+      maxAge: 
+        smallestMaxAge: 1
+        largestMaxAge: 31536000
+      preloadPolicy: "RequirePreload"
+      includeSubdomainsPolicy: "RequireIncludeSubdomains"
+```
+All routes that require HSTS will then use the same required Annotation on either cluster:
+````yaml
+apiVersion: v1
+kind: Route
+metadata:
+  annotations:
+    haproxy.router.openshift.io/hsts_header: max-age=31536000;preload;includeSubdomains
+...
+spec:
+  host: def.abc.com
+  tls:
+    termination: "reencrypt"
+    ...
+  wildcardPolicy: "Subdomain"
+````
+After applying the annotation, a route can be checked for the proper header like this:
+```shell
+$ curl -sik https://foo.com | grep strict-transport-security
+strict-transport-security: max-age=31536000
+````
+
+### User Stories
+
+#### As a cluster administrator, I want to verify HSTS globally, for all TLS routes in domain `foo.com`
```

Comment:
@soltysh have a think on what capabilities `oc` and/or our existing route fields selectors are missing to make the selection below easier to manage.

I think we're looking for
1. all routes with a particular annotation set
2. all routes with a particular unset
3. http/https field selector that can be used in conjunction with 1 or 2
4. a namespace label selector to constrain namespaces in conjunction with 3

This feature may be important enough to require an `oc routes hsts-something` command.  On the whole, I suspect the actions are predictable and the golang will be easier than bash.

Reply:
> * all routes with a particular annotation set
> * all routes with a particular unset

We don't have any annotation selectors, only label selectors, since that's what api server allows. Do we want to add that capability to oc/kubectl, it still will be client-side, though. 

> * http/https field selector that can be used in conjunction with 1 or 2

That should be possible already. 

> * a namespace label selector to constrain namespaces in conjunction with 3

We are iterating over namspaces for `-A` so we could theoretically filter out based on label selector, the question is only UX.